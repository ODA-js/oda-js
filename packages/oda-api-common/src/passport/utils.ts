import * as crypto from 'crypto';
import * as config from 'config';

export const hashPassword = password => {
  let salt = crypto.randomBytes(16).toString('base64');
  let hash = crypto
    .pbkdf2Sync(
      password,
      salt + config.get('passport.salt2'),
      10000,
      512,
      'sha512',
    )
    .toString('base64');
  return { salt, hash };
};

export const verifyPasswordWithSalt = (password, salt, hash) => {
  return (
    hash ===
    crypto
      .pbkdf2Sync(
        password,
        salt + config.get('passport.salt2'),
        10000,
        512,
        'sha512',
      )
      .toString('base64')
  );
};
