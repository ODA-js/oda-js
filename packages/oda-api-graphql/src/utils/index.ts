import get from './get';
import set from './set';
import fold from './fold';
import deepMerge from './deepMerge';
import padLeft from './padLeft';

export {
  set,
  get,
  fold,
  deepMerge,
  padLeft,
};
