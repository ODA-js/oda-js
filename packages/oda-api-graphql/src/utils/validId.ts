const checkForHexRegExp = new RegExp("^[0-9a-fA-F]{24}$");

export default function validId(id) {
  if (id == null) return false;

  if (typeof id == 'number') {
    return true;
  }

  if (typeof id == 'string') {
    return id.length == 12 || (id.length == 24 && checkForHexRegExp.test(id));
  }

  if (id.toHexString) {
    return id.id.length == 12 || (id.id.length == 24 && checkForHexRegExp.test(id.id));
  }

  if (id.toStirng) {
    return id.length == 12 || (id.length == 24 && checkForHexRegExp.test(id));
  }

  return false;
};