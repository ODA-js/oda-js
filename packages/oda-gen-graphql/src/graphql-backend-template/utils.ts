import { ModelPackage, FieldType } from 'oda-model';
import { template } from './model/packages.registerConnectors.ts';

export function decapitalize(name: string): string {
  return name[0].toLowerCase() + name.slice(1);
}

export function capitalize(name: string): string {
  return name[0].toUpperCase() + name.slice(1);
}

export function printRequired(field: { required?: boolean }): string {
  return field.required ? '!' : '';
}
// двух уровневая система распознавания типов
// 1. уровень парсит типы чтобы было понятно что за тип
// 2. уровень должен извлекать конкретные данные для конкретного типа, на основе полной информации о типе
// полная информация хранит первый тип и все выведенные типы....
//
// * means default type
export const defaultTypeMapper = {
  aor: {
    Number: ['int', 'integer', 'number', 'float', 'double'],
    Text: ['string', '*', 'uuid', 'id', 'identity'],
    Date: ['date', 'time', 'datetime'],
    Boolean: ['bool', 'boolean'],
    LongText: ['text'],
    RichText: ['richtext'],
    File: ['file'],
    Image: ['image'],
    'uix.components.$type': ['json'],
    'uix.enums.$type': ['enum()'],
  },
  resource: {
    number: ['int', 'integer', 'number', 'float', 'double'],
    string: ['string', 'text', '*', 'uuid', 'id', 'identity', 'richtext'],
    date: ['date', 'time', 'datetime'],
    boolean: ['bool', 'boolean'],
  },
  aorFilter: {
    Number: ['int', 'integer', 'number', 'float', 'double'],
    Text: ['string', 'text', '*', 'richtext'],
    ID: ['uuid', 'id', 'identity'],
    Date: ['date', 'time', 'datetime'],
    Boolean: ['bool', 'boolean'],
    [`#{partial('uix.enums.$type')}`]: ['enum()'],
  },
  graphql: {
    Int: ['int', 'integer'],
    Float: ['number', 'float', 'double'],
    String: ['string', 'text', '*', 'uuid', 'richtext'],
    JSON: ['object', 'json'],
    Date: ['date', 'time', 'datetime'],
    Boolean: ['bool', 'boolean'],
    ID: ['id', 'identity'],
    $type: ['entity()', 'enum()'],
    '[$type]': ['many()'],
  },
  mongoose: {
    Number: ['int', 'integer', 'number', 'float', 'double', 'identity'],
    String: ['string', 'text', '*', 'uuid', 'richtext'],
    'mongoose.Schema.Types.Mixed': ['object', 'json'],
    Boolean: ['bool', 'boolean'],
    Date: ['date'],
    'mongoose.Schema.Types.ObjectId': ['id'],
    '$type[]': ['many()'],
    'common.schema.$type[]': ['entity()'],
  },
  sequelize: {
    'DataTypes.INTEGER': ['int', 'integer', 'identity'],
    'DataTypes.FLOAT': ['number', 'float', 'double'],
    'DataTypes.STRING(1000)': ['string', 'text', '*', 'richtext'],
    'DataTypes.BOOLEAN': ['bool', 'boolean'],
    'DataTypes.DATE': ['date'],
    'DataTypes.CHAR(24)': ['id'],
    'DataTypes.INTEGER, autoIncrement: true': ['identity_pk'],
    'DataTypes.UUID': ['uuid'],
    'DataTypes.UUID, defaultValue: Sequelize.UUIDV4': ['uuid_pk'],
    'DataTypes.CHAR(24), defaultValue: ()=> IdGenerator.generateMongoId()': [
      'id_pk',
    ],
    'DataTypes.JSON': ['many()'],
    'DataTypes.STRING': ['enum()'],
  },
  typescript: {
    number: ['int', 'integer', 'number', 'float', 'double', 'identity'],
    string: ['string', 'text', 'id', '*', 'uuid', 'richtext'],
    boolean: ['bool', 'boolean'],
    Date: ['date'],
    object: ['json', 'object'],
    '$type[]': ['many()'],
    'common.$type': ['entity()', 'enum()'],
  },
};

export function prepareMapper(
  mapper: { [key: string]: string[] },
  systemPackage: ModelPackage,
) {
  const specificMapper = Object.keys(mapper).reduce((hash, current) => {
    mapper[current].forEach(t => {
      hash[t.toLowerCase()] = current;
    });
    return hash;
  }, {});

  const hasEnums = specificMapper.hasOwnProperty('enum()');
  const hasMany = specificMapper.hasOwnProperty('many()');
  const hasEntity = specificMapper.hasOwnProperty('entity()');
  return (type: FieldType | void) => {
    let result;
    if (typeof type === 'object' && type) {
      result = undefined;
      if (
        hasEntity &&
        type.type === 'entity' &&
        systemPackage.entities.has(type.name)
      ) {
        result = specificMapper['entity()'].replace(/\$type/gi, type.name);
      } else if (
        type.type === 'enum' &&
        hasEnums &&
        systemPackage.enums.has(type.name)
      ) {
        result = specificMapper['enum()'].replace(/\$type/gi, type.name);
      }
      if (!result) {
        result = specificMapper['*'];
      } else if (type.multiplicity === 'many' && hasMany) {
        result = specificMapper['many()'].replace(/\$type/gi, result);
      }
    } else if (typeof type === 'string') {
      result = specificMapper[type.toLowerCase()] || specificMapper['*'];
    } else {
      result = specificMapper['*'];
    }
    return result;
  };
}

export function printArguments(
  field: { args: any },
  typeMapper: (i: string) => string,
) {
  let result = '';
  if (field.args) {
    for (let arg of field.args) {
      let type = typeMapper(arg.type);
      result += `${arg.name}: ${type}${printRequired(arg)}`;
      if (arg.defaultValue) {
        if (type in { Int: 1, Float: 1, Boolean: 1 }) {
          result += ` = ${arg.defaultValue}`;
        } else {
          result += ` = "${arg.defaultValue}"`;
        }
      }
    }
  }
  return result;
}
