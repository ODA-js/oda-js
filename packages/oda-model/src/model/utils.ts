import { Map, Set } from 'immutable';

import { IFieldContext } from '../contexts/IFieldContext';
import { IMutationContext } from '../contexts/IMutationContext';
import { IRelationContext } from '../contexts/IRelationContext';
import { IEntity } from '../interfaces/IEntity';
import { IEntityRef } from '../interfaces/IEntityRef';
import { IField, IFieldInit } from '../interfaces/IField';
import { FieldArgsInput, IFieldArg, IFieldArgInit } from '../interfaces/IFieldArg';
import { IMutation } from '../interfaces/IMutation';
import { IRelation } from '../interfaces/IRelation';
import { EntityRef } from './EntityRef';
import { ModelFactory } from './Factory';
import { Field } from './Field';
import { FieldArg } from './FieldArg';
import { FieldArgsTransform, EntityRefTransform, FieldTransformType } from '../interfaces/types';

export type ArrayToSet<S> = {
  transform: (input: S[]) => Set<S>;
  reverse: (input: Set<S>) => S[];
};

export function TransformRef(): EntityRefTransform {
  return {
    transform: (inp: string | IEntityRef, relation: IRelation) => {
      if (inp) {
        return new EntityRef(inp, ModelFactory.getContext(relation) as IRelationContext);
      } else {
        return null;
      }
    },
    reverse: (inp: IEntityRef) => {
      if (inp) {
        return inp.toString();
      } else {
        return null;
      }
    },
  };
}

export function TransformField(): FieldTransformType {
  return {
    transform: (input: {
      [name: string]: Partial<IFieldInit>,
    } | IFieldInit[], owner: IEntity | IRelation) => {
      const context = ModelFactory.getContext(owner) as IRelationContext;
      if (!Array.isArray(input)) {
        input = Object.keys(input).map(k => ({
          name: k,
          ...input[k],
        }));
      }
      return Map<string, IField>(input.map(p => [p.name, new Field(p, context)]) as [string, IField][]);
    },
    reverse: (input: Map<string, IField>) => {
      if (input) {
        return Array.from(input.values()[Symbol.iterator]()).map(i => i.toJS() as IFieldInit);
      } else {
        return null;
      }
    },
  };
}

export function TransformArgs(): FieldArgsTransform {
  return {
    transform: (input: FieldArgsInput, owner: IMutation | IField ) => {
      if (input) {
        const context = ModelFactory.getContext(owner) as IMutationContext | IFieldContext;
        if (!Array.isArray(input)) {
          input = Object.keys(input).map(k => ({
            name: k,
            ...input[k],
          }));
        }
        return Map<string, IFieldArg>(input.map(p => [p.name, new FieldArg(p, context)]) as [string, IFieldArg][]);
      } else {
        return null;
      }
    },
    reverse: (input: Map<string, IFieldArg>) => {
      if (input) {
        return Array.from(input.values()[Symbol.iterator]()).map(p => p.toJS() as IFieldArgInit) ;
      } else {
        return null;
      }
    },
  };
}
