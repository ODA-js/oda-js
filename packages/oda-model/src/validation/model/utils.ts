import { Map, Set } from 'immutable';

export type MapType<T, S> = {
  transform: (input: T) => S;
  reverse: (input: S) => T;
};

export type ArrayToMap<S extends object> = {
  transform: (input: S[]) => Map<string, S>;
  reverse: (input: Map<string, S>) => S[];
};

export type ArrayToSet<S> = {
  transform: (input: S[]) => Set<S>;
  reverse: (input: Set<S>) => S[];
};

export function transformMap<S extends { name: string }>(): ArrayToMap<S> {
  return {
    transform: (input: S[]) => Map<string, S>(input.map(p => [p.name, p]) as [string, S][]),
    reverse: (input: Map<string, S>) => Array.from(input.values()[Symbol.iterator]()),
  };
}


export function transformSet<S>(): ArrayToSet<S> {
  return {
    transform: (input: S[]) => Set<S>(input),
    reverse: (input: Set<S>) => Array.from(input.values()[Symbol.iterator]()),
  };
}
