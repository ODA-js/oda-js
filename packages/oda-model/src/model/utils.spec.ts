import { TransformArgs } from './utils';

describe('transform Fields', () => {
  let transform = TransformArgs();
  it('transform', () => {
    const result = transform.transform([
      { name: 'one'},
      { name: 'two', required: true },
      { name: 'three', required: false, defaultValue: '3' },
      { name: 'four', required: false, defaultValue: '3', type: 'Number' },
    ]);
    expect(result.has('one'));
    expect(result.has('two'));
    expect(result.has('three'));
    expect(result.has('four'));
    expect(result.get('one').defaultValue).toBeNull();
    expect(result.get('one').required).toBeNull();
    expect(result.get('one').type).toBeNull();
  });
});
