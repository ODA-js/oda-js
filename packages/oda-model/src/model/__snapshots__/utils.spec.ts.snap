// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform Fields transform 1`] = `
Immutable.Map {
  "one": FieldArg {
    "store": Immutable.Record {
      "name": "one",
      "title": null,
      "description": null,
      "required": null,
      "type": null,
      "defaultValue": null,
    },
  },
  "two": FieldArg {
    "store": Immutable.Record {
      "name": "two",
      "title": null,
      "description": null,
      "required": true,
      "type": null,
      "defaultValue": null,
    },
  },
  "three": FieldArg {
    "store": Immutable.Record {
      "name": "three",
      "title": null,
      "description": null,
      "required": false,
      "type": null,
      "defaultValue": "3",
    },
  },
  "four": FieldArg {
    "store": Immutable.Record {
      "name": "four",
      "title": null,
      "description": null,
      "required": false,
      "type": "Number",
      "defaultValue": "3",
    },
  },
}
`;

exports[`transform Fields transform 2`] = `
Array [
  Object {
    "name": "one",
  },
  Object {
    "name": "two",
    "required": true,
  },
  Object {
    "defaultValue": "3",
    "name": "three",
    "required": false,
  },
  Object {
    "defaultValue": "3",
    "name": "four",
    "required": false,
    "type": "Number",
  },
]
`;

exports[`transform Fields transform hash 1`] = `
Immutable.Map {
  "one": FieldArg {
    "store": Immutable.Record {
      "name": "one",
      "title": null,
      "description": null,
      "required": null,
      "type": null,
      "defaultValue": null,
    },
  },
  "two": FieldArg {
    "store": Immutable.Record {
      "name": "two",
      "title": null,
      "description": null,
      "required": true,
      "type": null,
      "defaultValue": null,
    },
  },
  "three": FieldArg {
    "store": Immutable.Record {
      "name": "three",
      "title": null,
      "description": null,
      "required": false,
      "type": null,
      "defaultValue": "3",
    },
  },
  "four": FieldArg {
    "store": Immutable.Record {
      "name": "four",
      "title": null,
      "description": null,
      "required": false,
      "type": "Number",
      "defaultValue": "3",
    },
  },
}
`;

exports[`transform Fields transform hash 2`] = `
Array [
  Object {
    "name": "one",
  },
  Object {
    "name": "two",
    "required": true,
  },
  Object {
    "defaultValue": "3",
    "name": "three",
    "required": false,
  },
  Object {
    "defaultValue": "3",
    "name": "four",
    "required": false,
    "type": "Number",
  },
]
`;

exports[`transform Ref transform 1`] = `null`;

exports[`transform Ref transform 2`] = `null`;

exports[`transform Ref transform 3`] = `null`;
